### PR TITLE
Optimise locking in rsa_get_blinding()

### LINKED ISSUE #20286: 3.0 performance degraded due to locking
Using some internal QUIC tests, we see this timing in 1.1:
```
$ USE_GQUIC_VERSIONS=1 perf record -F 999 -g ./quic_lib_tests  --gtest_repeat=100  --gtest_filter=*ZeroRttDisabled*
$ perf report

+   31.73%    31.55%  quic_lib_tests  quic_lib_tests       [.] bn_sqr8x_internal
+    9.28%     9.28%  quic_lib_tests  quic_lib_tests       [.] mul4x_internal
+    4.91%     4.91%  quic_lib_tests  quic_lib_tests       [.] sha256_block_data_order_avx
```

In 3.0 we see this:
```

$ USE_GQUIC_VERSIONS=1 perf record -F 999 -g ./quic_lib_tests  --gtest_repeat=100  --gtest_filter=*ZeroRttDisabled*
$ perf report

+   11.02%    10.99%  quic_lib_tests  quic_lib_tests       [.] bn_sqr8x_internal
+    8.38%     8.08%  quic_lib_tests  libpthread-2.31.so   [.] __pthread_rwlock_rdlock
+    7.65%     7.51%  quic_lib_tests  libpthread-2.31.so   [.] __pthread_rwlock_unlock
+    4.98%     4.78%  quic_lib_tests  quic_lib_tests       [.] getrn
+    4.14%     4.11%  quic_lib_tests  quic_lib_tests       [.] mul4x_internal
+    3.37%     2.57%  quic_lib_tests  quic_lib_tests       [.] ossl_tolower
     3.30%     3.30%  quic_lib_tests  quic_lib_tests       [.] ossl_lh_strcasehash
+    2.72%     2.13%  quic_lib_tests  quic_lib_tests       [.] OPENSSL_strcasecmp
+    2.29%     2.05%  quic_lib_tests  quic_lib_tests       [.] ossl_lib_ctx_get_data
+    1.93%     1.93%  quic_lib_tests  quic_lib_tests       [.] sha256_block_data_order_avx
```

This seems to be part of `OSSL_DECODER_CTX_new_for_pkey`  16% of the time is spent in locking, on a *single threaded binary.* And 10% is in a string hashtable lookup.

If anyone on the project is going to look at this, I will try to get a small reproducer. But our the time for our QUIC tests is doubling.

### PR DESCRIPTION
**Pull request by mattcaswell**
We optimise locking in rsa_get_blinding() so that we normally take a read lock, and only fallback to a write lock if we need to. This will be very slightly slower in the case of single use RSA objects, but should be significantly better when an RSA object is reused in a multi-threaded environment. It's probably worth the trade off.

Partially fixes #20286

### CODE REVIEW

File: crypto/rsa/rsa_ossl.c [hunk 1]
```diff
@@ -186,11 +186,21 @@ static BN_BLINDING *rsa_get_blinding(RSA *rsa, int *local, BN_CTX *ctx)
...
 
-    if (!CRYPTO_THREAD_write_lock(rsa->lock))
+    if (!CRYPTO_THREAD_read_lock(rsa->lock))
         return NULL;
 
     if (rsa->blinding == NULL) {
-        rsa->blinding = RSA_setup_blinding(rsa, ctx);
+        /*
+         * This dance with upgrading the lock from read to write will be
+         * slower in cases of a single use RSA object, but should be
```
Comment (paulidale) [inline crypto/rsa/rsa_ossl.c:195]: There shouldn't be any contention for a single use object, so the cost ought to be trivial compared to the operation.

File: crypto/rsa/rsa_ossl.c [hunk 2]
```diff
@@ -186,11 +186,21 @@ static BN_BLINDING *rsa_get_blinding(RSA *rsa, int *local, BN_CTX *ctx)
...
+         * significantly better in multi-thread cases (e.g. servers). It's
+         * probably worth it.
+         */
+        CRYPTO_THREAD_unlock(rsa->lock);
+        if (!CRYPTO_THREAD_write_lock(rsa->lock))
+            return NULL;
+        if (rsa->blinding == NULL)
```
Comment (paulidale) [inline crypto/rsa/rsa_ossl.c:202]: This isn't required, it's already been established that ->blinding is NULL.

File: crypto/rsa/rsa_ossl.c [hunk 3]
```diff
@@ -212,7 +222,11 @@ static BN_BLINDING *rsa_get_blinding(RSA *rsa, int *local, BN_CTX *ctx)
         *local = 0;
 
         if (rsa->mt_blinding == NULL) {
-            rsa->mt_blinding = RSA_setup_blinding(rsa, ctx);
+            CRYPTO_THREAD_unlock(rsa->lock);
+            if (!CRYPTO_THREAD_write_lock(rsa->lock))
+                return NULL;
+            if (rsa->mt_blinding == NULL)
```
Comment (paulidale) [inline crypto/rsa/rsa_ossl.c:228]: Same comment here ->mt_blinding is NULL at this point.

**Review by paulidale — APPROVED** [review]
Two amendments and then good.

**Review by t8m — APPROVED** [review]

**Comment by mattcaswell** [comment]
I've run some performance tests on this patch based on the test in openssl/tools#152

The test runs EVP_PKEY_sign_init()/EVP_PKEY_sign() repeatedly in a loop using a 512-bit RSA key. I ran the test against different numbers of threads (1, 10, 100, 500, 1000, 2000) and against different OpenSSL versions (1.1.1, 3.0, 3.1, master, master + this patch).

The results are in us and show the average time to execute 100 EVP_PKEY_sign_init()/EVP_PKEY_sign() operations. Smaller numbers are better.

  | 1 | 10 | 100 | 500 | 1000 | 2000
-- | -- | -- | -- | -- | -- | --
  | 5922.47 | 1118.766 | 936.8872 | 1003.88558 | 989.82089 | 1072.68209
3.0 | 5942.49 | 1107.251 | 894.7091 | 981.5546 | 1106.35009 | 1108.78731
3.1 | 6030.36 | 1113.128 | 903.1216 | 998.79938 | 972.42778 | 1043.656205
master | 6024.82 | 1054.014 | 870.7651 | 925.0781 | 979.64881 | 1035.024995
master + #20953 | 6095.77 | 961.017 | 875.3859 | 944.98296 | 1014.76314 | 1090.16631

This does not seem to indicate any measurable performance benefit from this patch. This is most likely because the time spent in the RSA sign operation per loop is quite long, and therefore the likelihood of any lock contention is reduced.

**Comment by t8m** [comment]
**OTC: master only as there is no significant performance benefit.**

**Comment by mattcaswell** [comment]
Pushed to master. Thanks.

**Pull request closed without merging**